src/zx/FunctionalityConstruction.cpp

```diff
@@ -0,0 +1,425 @@
+        //     std::vector<Vertex> new_qubit_vertices(qc->getNqubits());
+        //     for (auto i = 0; i < qc->getNqubits(); i++) {
+        //         new_qubit_vertices[i] = qubit_vertices[i];
+        //     }
+        //     for (auto& [tar, src]:
+        //          qc->initialLayout) { // reverse initial permutation
+        //         if (tar == src)
+        //             continue;
+
+        //         auto v_tar = diag.addVertex(tar, 1);
+        //         diag.addEdge(qubit_vertices[src], v_tar);
+        //         new_qubit_vertices[tar] = v_tar;
+        //     }
+        //     qubit_vertices = new_qubit_vertices;
+        // }
+
+        for (auto it = qc->cbegin(); it != qc->cend();) {
+            auto& op = *it;
+
+            if (op->getType() == qc::OpType::Compound) {
+                auto* compOp = dynamic_cast<qc::CompoundOperation*>(op.get());
+                for (auto subIt = compOp->cbegin(); subIt != compOp->cend();)
+                    subIt = parse_op(diag, subIt, compOp->end(), qubit_vertices, qc->initialLayout);
+                ++it;
+            } else {
+                it = parse_op(diag, it, qc->end(), qubit_vertices, qc->initialLayout);
+            }
+        }
+
+        for (size_t i = 0; i < qubit_vertices.size(); ++i) {
+            diag.addEdge(qubit_vertices[i], diag.getOutputs()[i]);
+        }
+        return diag;
+    }
+    bool FunctionalityConstruction::transformableToZX(const qc::QuantumComputation* qc) {
+        for (const auto& it: *qc) {
+            if (!transformableToZX(it.get()))
+                return false;
+        }
+        return true;
+    }
+
+    bool FunctionalityConstruction::transformableToZX(qc::Operation* op) {
+        if (op->getType() == qc::OpType::Compound) {
+            auto* compOp = dynamic_cast<qc::CompoundOperation*>(op);
+
+            for (const auto& it: *compOp) {
+                if (!transformableToZX(it.get()))
+                    return false;
+            }
+            return true;
+        }
+
+        if (op->getType() == qc::OpType::Barrier) {
+            return true;
+        }
+
+        if (!op->isControlled()) {
+            switch (op->getType()) {
+                case qc::OpType::Z:
+
+                case qc::OpType::RZ:
+                case qc::OpType::Phase:
+                case qc::OpType::X:
+                case qc::OpType::RX:
+                case qc::OpType::Y:
+                case qc::OpType::RY:
+                case qc::OpType::T:
+                case qc::OpType::Tdag:
+                case qc::OpType::S:
+                case qc::OpType::Sdag:
+                case qc::OpType::U2:
+                case qc::OpType::U3:
+                case qc::OpType::SWAP:
+                case qc::OpType::H:
+                case qc::OpType::Measure:
+                case qc::OpType::I: {
+                    return true;
+                }
+                default: {
+                    return false;
+                }
```

Comment:
switch statements typically do not need their own code block and I believe it to be best practice to not use them if they are not necessary (as is the case when you want to have temporary variables in one particular `case`)